db-engine/src/main/java/com/splicemachine/db/catalog/types/IndexDescriptorImpl.java

```diff
@@ -553,10 +564,90 @@ public BaseExecutableIndexExpression getExecutableIndexExpression(int indexColum
                 : "index has expression but generated class name is unknown";
         LanguageConnectionContext lcc = (LanguageConnectionContext) ContextService.getContext
                 (LanguageConnectionContext.CONTEXT_ID);
+        assert lcc != null;
         ClassFactory classFactory = lcc.getLanguageConnectionFactory().getClassFactory();
         GeneratedClass gc = classFactory.loadGeneratedClass(
                 generatedClassNames[indexColumnPosition], exprBytecode[indexColumnPosition]);
         executableExprs[indexColumnPosition] = (BaseExecutableIndexExpression) gc.newInstance(lcc);
         return executableExprs[indexColumnPosition];
     }
+
+    /** @see IndexDescriptor#getParsedIndexExpressions */
+    @Override
+    public ValueNode[] getParsedIndexExpressions(LanguageConnectionContext context, Optimizable optTable)
+            throws StandardException
+    {
+        LanguageConnectionContext lcc = context;
+        if (lcc == null) {
+            lcc = (LanguageConnectionContext) ContextService.getContext(LanguageConnectionContext.CONTEXT_ID);
+        }
+        assert lcc != null;
+
+        if (exprAsts != null) {
+            for (ValueNode ast : exprAsts) {
+                setTableNumberToIndexExpr(ast, optTable);
+                bindNecessaryNodesInIndexExpr(ast, optTable, lcc);
```

Comment:
@ascend1 Could you check if IndexDescriptorImpl is cached as part of the TableDescriptor in dictionary cache? If so, it is shared among all the queries and the these code could cause the data structure to be modified concurrently, and that could be problematic.

Reply:
@yxia92 I looked into the code again and I think `IndexDescriptorImpl` will be effectively cached because it's part of `ConglomerateDescriptor`, and `TableDescriptor` has a `conglomerateDescriptorList`. Thanks for pointing this out, it's indeed a serious problem.

For this change, I can remove the cache and let the method always parse. I mean, at this point, it doesn't seem to be a bad idea because it's correct and simple. Options to improve it could be:

1. Cache the ASTs in a statement-level cache so that different queries do not step on each other's feet. I'm not sure which objects' life cycle align with a statement and are always accessible during optimization. Maybe `CompilerContext`? It's managed in a stack, doesn't seem to be per statement. `LCC`? It's probably alive for the whole session?
2. Implement `clone()` method to `ValueNode` class hierarchy and always return a deep copy of ASTs. IMHO, code would be more readable and maintainable in this way. Also, `clone()` is needed (I encountered two places) by some logic and currently we cannot do that for index expressions. But then back to the beginning, this probably has only marginal performance improvement comparing to always parsing since in both way we need to construct the same amount of objects. Only the grammar matching of parsing is saved, which is probably not much anyway, since javacc doesn't do backtracking and index expressions are usually short.

Reply:
Thanks @ascend1 for checking! I'm good with moving this change out from the current PR. CompilerContext as I understand is per-query, and it should be like a global structure visible during the query compilation phase for a particular query.